### ISSUE TITLE
New method: Sequence.Traverse

### ISSUE THREAD

**Issue by atifaziz** [morelinq/MoreLINQ#14]
`Sequence.Generate` is able to traverse linked lists of reference types, but it produces an infinite list and also doesn't check for null (nor should/could it, generally speaking).

So it has to be used in a boilerplate idiom where it is combined with `TakeWhile`, e.g.

```c#
Exception innermost = Sequence.Generate(caught, e => e.InnerException).While(e => e != null).Last();
```

The name Generate then becomes a little misleading, as we are not generating a new list but traversing an existing list.

Also it is a bit counter-intuitive to have the check for null termination appearing after the `Generate` call - perfectly safe due to the laziness, but it looks dodgy.

Given the usefulness of this pattern, and the fact that we are not really generating anything, it might make sense to make it a separate method called `Traverse`, limited of course to reference types:

```c#
public static IEnumerable<TResult> Traverse<TResult>(TResult initial,
    Func<TResult, TResult> generator) where TResult : class
{
    for (TResult current = initial; current != null; current = generator(current))
        yield return current;
}
```

---
Originally reported on Google Code with ID 14

Reported by `daniel.earwicker` on 2009-03-10 11:18:24

**Comment by leandromoh** [comment]
As already mentioned by CAmmerman, `unfold `function can solve it;

`Exception innermost = Unfold(caught, e => e.InnerException != null ? Tuple.Create(e, e.InnerException) : null).Last();`

**Comment by leandromoh** [comment]
@atifaziz will you close this issue now that `unfold `were implemented ?

**Comment by atifaziz** [comment]
@leandromoh Yep & thanks for helping & remembering to triage this.

**Comment by atifaziz** [comment]
In [final form `Unfold` was released in 2.3](https://morelinq.github.io/2.3/ref/api/html/M_MoreLinq_MoreEnumerable_Unfold__3.htm), the example of traversing nested exceptions to find the innermost will look like this:

```c#
var innermost = 
    MoreEnumerable.Unfold(caught, e => e, e => e != null, e => e.InnerException, e => e)
                  .Last();
```
